File: smoke_tests/swift/path_traversal.swift
// Path Traversal vulnerabilities in Swift
import Foundation

class PathTraversalVulnerabilities {
    func readFileUnsafe(filename: String) throws -> String {
        // VULNERABLE: Path traversal via concatenation
        let path = "/var/data/\(filename)"
        return try String(contentsOfFile: path, encoding: .utf8)
    }

    func serveFileUnsafe(userPath: String) throws -> Data {
        // VULNERABLE: No path validation
        let url = URL(fileURLWithPath: "/public/files/\(userPath)")
        return try Data(contentsOf: url)
    }

    func deleteFileUnsafe(filename: String) throws {
        // VULNERABLE: Arbitrary file deletion
        let path = "/tmp/\(filename)"
        try FileManager.default.removeItem(atPath: path)
    }

    func writeFileUnsafe(filename: String, content: String) throws {
        // VULNERABLE: User controls path
        let path = "/uploads/\(filename)"
        try content.write(toFile: path, atomically: true, encoding: .utf8)
    }

    func listDirectoryUnsafe(dirName: String) throws -> [String] {
        // VULNERABLE: Directory listing with user input
        let path = "/data/\(dirName)"
        return try FileManager.default.contentsOfDirectory(atPath: path)
    }
}

File: smoke_tests/php/path_traversal.php
<?php
// Path Traversal vulnerabilities in PHP

class PathTraversalVulnerabilities {
    public function readFileUnsafe($filename) {
        // VULNERABLE: Path traversal via concatenation
        $path = '/var/data/' . $filename;
        return file_get_contents($path);
    }

    public function includeFileUnsafe($page) {
        // VULNERABLE: LFI via include
        include('/templates/' . $page . '.php');
    }

    public function requireFileUnsafe($module) {
        // VULNERABLE: LFI via require
        require($module);
    }

    public function deleteFileUnsafe($filename) {
        // VULNERABLE: Arbitrary file deletion
        $path = '/tmp/' . $filename;
        unlink($path);
    }

    public function readFileUrl($url) {
        // VULNERABLE: Remote file inclusion
        return file_get_contents($url);
    }

    public function downloadFile() {
        // VULNERABLE: Path traversal in download
        $file = $_GET['file'];
        header('Content-Disposition: attachment; filename="' . basename($file) . '"');
        readfile('/downloads/' . $file);
    }

    public function writeFileUnsafe($filename, $content) {
        // VULNERABLE: Arbitrary file write
        file_put_contents('/uploads/' . $filename, $content);
    }
}
